### PR TITLE
Provide more information for HRTB lifetime errors involving closures

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: src/test/ui/lifetimes/issue-79187-2.stderr [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+note: this closure does not fulfill the lifetime requirements
+  --> $DIR/issue-79187-2.rs:8:14
+   |
+LL |     take_foo(|a| a);
+   |              ^^^^^
+note: the lifetime requirement is introduced here
+  --> $DIR/issue-79187-2.rs:5:21
+   |
+LL | fn take_foo(_: impl Foo) {}
+   |                     ^^^
```
Comment (estebank) [inline src/test/ui/lifetimes/issue-79187-2.stderr:18]: I hadn't considered this case in particular, but ideally we would (also) point at the `where` clause in `impl<F> Foo for F where F: Fn(&i32) -> &i32` for this...

**Comment by estebank** [comment]
This is a mitigation of a stable to stable regression (https://github.com/rust-lang/rust/issues/75791), but not a fix.

File: src/test/ui/lifetimes/issue-79187-2.nll.stderr [hunk 2]
```diff
@@ -0,0 +1,44 @@
+error: lifetime may not live long enough
+  --> $DIR/issue-79187-2.rs:9:24
+   |
+LL |     take_foo(|a: &i32| a);
+   |                  -   - ^ returning this value requires that `'1` must outlive `'2`
+   |                  |   |
+   |                  |   return type of closure is &'2 i32
+   |                  let's call the lifetime of this reference `'1`
```
Comment (oli-obk) [inline src/test/ui/lifetimes/issue-79187-2.nll.stderr:8]: Not sure if we have the information available anywhere, but could emit a diagnostic of this quality for the `the required lifetime does not necessarily outlive the anonymous lifetime #1 defined on the body`, too?

File: src/test/ui/lifetimes/issue-79187-2.nll.stderr [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+error: lifetime may not live long enough
+  --> $DIR/issue-79187-2.rs:10:34
+   |
+LL |     take_foo(|a: &i32| -> &i32 { a });
+   |                  -        -      ^ returning this value requires that `'1` must outlive `'2`
+   |                  |        |
+   |                  |        let's call the lifetime of this reference `'2`
+   |                  let's call the lifetime of this reference `'1`
+
+error: higher-ranked subtype error
```
Comment (oli-obk) [inline src/test/ui/lifetimes/issue-79187-2.nll.stderr:19]: :eyes: oh wow, talking about ~~high quality~~ errors. I don't think I've ever encountered this error

File: src/test/ui/lifetimes/issue-79187.rs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+fn thing(x: impl FnOnce(&u32)) {}
+
+fn main() {
+    let f = |_| ();
+    thing(f); //~ERROR mismatched types
+}
```
Comment (oli-obk) [inline src/test/ui/lifetimes/issue-79187.rs:6]: missing trailing newline

**Comment by oli-obk** [comment]
The test diffs lgtm, but some of these new tests you added highlight some seriously unhelpful errors. Most of them have an independently emitted helpful error, but not all

File: src/test/ui/lifetimes/issue-79187-2.nll.stderr (hunk 3 above)
Comment (estebank) [reply src/test/ui/lifetimes/issue-79187-2.nll.stderr:19]: I'm not surprised, it's NLL only :-/

**Comment by estebank** [comment]
> The test diffs lgtm, but some of these new tests you added highlight some seriously unhelpful errors. Most of them have an independently emitted helpful error, but not all

Indeed. This is gonna be a rabbit-hole to clean up.

**Comment by oli-obk** [comment]
r=me with CI passing (probably just the missing newline). Let's address further improvements in other PRs, it's not necessary to do more in this one

**Pull request merged**